com.squareup:javawriter included but not declared

I see you included com.squareup:javawriter in the jar package. This leads to dupplicate symbols if another dependency properly declares com.squareup:javawriter as a dependency.

For instance espresso 2.0 relies on javawriter but it is properly declared in their pom.

I´ve prepared a sample project but I cannot reproduce the exact problem. I´ve copied our list of depdendencies.

Here is the sample project: https://github.com/falkorichter/realm-javawriter-dup-
Here is the error I get:
```
:library:compileDebugTestJava                 
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
:library:packageAllDebugTestClassesForMultiDex FAILED          
              
FAILURE: Build failed with an exception.
              
* What went wrong:
Execution failed for task ':library:packageAllDebugTestClassesForMultiDex'.
> java.util.zip.ZipException: duplicate entry: com/squareup/javawriter/JavaWriter$Scope.class
```              

Espresso has JavaWriter as a declared dependency (output of ./gradlew dependencies):
```
androidTestCompile - Classpath for compiling the androidTest sources.
+--- com.squareup:fest-android:1.0.8                                                        
|    +--- com.android.support:support-v4:19.1.+ -> 19.1.0
|    \--- org.easytesting:fest-assert-core:2.0M10
|         \--- org.easytesting:fest-util:1.2.5
+--- org.mockito:mockito-core:1.9.5  
|    +--- org.hamcrest:hamcrest-core:1.1
|    \--- org.objenesis:objenesis:1.0
+--- com.google.dexmaker:dexmaker:1.0
+--- com.google.dexmaker:dexmaker-mockito:1.0
|    +--- com.google.dexmaker:dexmaker:1.0
|    \--- org.mockito:mockito-core:1.9.5 (*)
+--- com.android.support.test.espresso:espresso-core:2.0
|    +--- com.squareup:javawriter:2.1.1
|    +--- org.hamcrest:hamcrest-integration:1.1
|    |    \--- org.hamcrest:hamcrest-core:1.1
|    +--- org.hamcrest:hamcrest-library:1.1
|    |    \--- org.hamcrest:hamcrest-core:1.1
|    +--- javax.inject:javax.inject:1
|    +--- com.android.support.test.espresso:espresso-idling-resource:2.0
|    +--- com.android.support.test:testing-support-lib:0.1
|    |    \--- junit:junit-dep:4.10  
|    |         \--- org.hamcrest:hamcrest-core:1.1
|    +--- com.google.code.findbugs:jsr305:2.0.1
|    +--- javax.annotation:javax.annotation-api:1.2
|    \--- org.hamcrest:hamcrest-core:1.1
+--- com.android.support.test:testing-support-lib:0.1 (*)
+--- org.apache.commons:commons-io:1.3.2
|    \--- commons-io:commons-io:1.3.2
\--- com.squareup.okhttp:mockwebserver:2.1.0
     \--- org.bouncycastle:bcprov-jdk15on:1.50
```

When unpacking your jar I see you included the Jarwriter without changing the package. This will lead to colisions.  I found https://github.com/realm/realm-java/pull/348 as a releated issue.

--- a/realm-annotations-processor/build.gradle
+++ b/realm-annotations-processor/build.gradle
@@ -1,4 +1,14 @@
+buildscript {
+    repositories {
+        jcenter()
+    }
+    dependencies {
+        classpath 'com.github.jengelman.gradle.plugins:shadow:1.2.1'
+    }
+}
+
 apply plugin: 'java'
+apply plugin: 'com.github.johnrengelman.shadow'
 
 version = new File("${projectDir}/../version.txt").text
 sourceCompatibility = '1.6'
@@ -26,6 +36,10 @@ dependencies {
 
 jar {
     from configurations.compile.findAll {it.name.contains('javawriter')}.collect { zipTree(it) }
+}
+
+shadowJar {
+    relocate 'com.squareup.javawriter', 'io.realm.processor.javawriter'
 }
 
 //for Ant filter

--- a/realm-annotations-processor/build.gradle
+++ b/realm-annotations-processor/build.gradle
@@ -40,6 +40,8 @@ jar {
 
 shadowJar {
     relocate 'com.squareup.javawriter', 'io.realm.processor.javawriter'
+    exclude 'io/realm/annotations/**'
+    classifier = ''
 }
 
 //for Ant filter

--- a/realm-annotations-processor/build.gradle
+++ b/realm-annotations-processor/build.gradle
@@ -74,7 +74,7 @@ task assembleRealm(type: GradleBuild) {
     tasks = ['assemble']
 }
 
-tasks.assemble {
+tasks.compileJava {
     dependsOn compileAnnotations
 }
 

--- a/realm/build.gradle
+++ b/realm/build.gradle
@@ -28,7 +28,7 @@ android {
 dependencies {
     compile 'com.intellij:annotations:+@jar'
     compile files("../realm-annotations/build/libs/realm-annotations-${version}.jar")
-    androidTestApt files("../realm-annotations-processor/build/libs/realm-annotations-processor-${version}-all.jar")
+    androidTestApt files("../realm-annotations-processor/build/libs/realm-annotations-processor-${version}.jar")
     androidTestApt files("../realm-annotations/build/libs/realm-annotations-${version}.jar")
 }
 
@@ -63,7 +63,7 @@ task androidJar(type: Jar, dependsOn: ['assemble']) {
     group 'Build'
     description 'Generates a jar file continaining Realm and its annotation processor'
     from zipTree('build/intermediates/bundles/release/classes.jar')
-    from zipTree("../realm-annotations-processor/build/libs/realm-annotations-processor-${version}-all.jar")
+    from zipTree("../realm-annotations-processor/build/libs/realm-annotations-processor-${version}.jar")
     from zipTree("../realm-annotations/build/libs/realm-annotations-${version}.jar")
     from(file('src/main/jniLibs')) {
         into 'lib'

--- a/realm/build.gradle
+++ b/realm/build.gradle
@@ -28,7 +28,7 @@ android {
 dependencies {
     compile 'com.intellij:annotations:+@jar'
     compile files("../realm-annotations/build/libs/realm-annotations-${version}.jar")
-    androidTestApt files("../realm-annotations-processor/build/libs/realm-annotations-processor-${version}.jar")
+    androidTestApt files("../realm-annotations-processor/build/libs/realm-annotations-processor-${version}-all.jar")
     androidTestApt files("../realm-annotations/build/libs/realm-annotations-${version}.jar")
 }
 
@@ -63,7 +63,7 @@ task androidJar(type: Jar, dependsOn: ['assemble']) {
     group 'Build'
     description 'Generates a jar file continaining Realm and its annotation processor'
     from zipTree('build/intermediates/bundles/release/classes.jar')
-    from zipTree("../realm-annotations-processor/build/libs/realm-annotations-processor-${version}.jar")
+    from zipTree("../realm-annotations-processor/build/libs/realm-annotations-processor-${version}-all.jar")
     from zipTree("../realm-annotations/build/libs/realm-annotations-${version}.jar")
     from(file('src/main/jniLibs')) {
         into 'lib'

--- a/realm/build.gradle
+++ b/realm/build.gradle
@@ -143,7 +143,7 @@ if (version.endsWith('-SNAPSHOT')) { // Only publish if it's a snapshot version
 
 task buildApt(type: GradleBuild) {
     dir = file('../realm-annotations-processor')
-    tasks = ['assemble']
+    tasks = ['shadowJar']
 }
 
 task compileJni(type: GradleBuild) {

